airflow/contrib/hooks/gcp_compute_hook.py

```diff
@@ -0,0 +1,122 @@
+# -*- coding: utf-8 -*-
+#
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+import time
+from googleapiclient.discovery import build
+
+from airflow import AirflowException
+from airflow.contrib.hooks.gcp_api_base_hook import GoogleCloudBaseHook
+
+# Number of retries - used by googleapiclient method calls to perform retries
+# For requests that are "retriable"
+NUM_RETRIES = 5
```

Comment:
Is this retrying for a longrunning operation (e.g. polling until it is done)? Or retrying in the event of 500 errors? I think the latter, but might want to clarify that in the comment.

Reply:
The retry logic is handled by the googleapiclient's `http.py`. We only pass it to `execute()` here.

Accourding to the [source code](https://chromium.googlesource.com/external/github.com/google/google-api-python-client/+/refs/heads/master/googleapiclient/http.py#795) the retry can happen in multiple cases. Below is the relevant code snippet:

```
def _should_retry_response(resp_status, content):
    """Determines whether a response should be retried.
    Args:
      resp_status: The response status received.
      content: The response content body.
    Returns:
      True if the response should be retried, otherwise False.
    """
    # Retry on 5xx errors.
    if resp_status >= 500:
        return True
    # Retry on 429 errors.
    if resp_status == _TOO_MANY_REQUESTS:
        return True
    # For 403 errors, we have to check for the `reason` in the response to
    # determine if we should retry.
    if resp_status == six.moves.http_client.FORBIDDEN:
        # If there's no details about the 403 type, don't retry.
        if not content:
            return False
        # Content is in JSON format.
        try:
            data = json.loads(content.decode('utf-8'))
            if isinstance(data, dict):
                reason = data['error']['errors'][0]['reason']
            else:
                reason = data[0]['error']['errors']['reason']
        except (UnicodeDecodeError, ValueError, KeyError):
            LOGGER.warning('Invalid JSON content from response: %s', content)
            return False
        LOGGER.warning('Encountered 403 Forbidden with reason "%s"', reason)
        # Only retry on rate limit related failures.
        if reason in ('userRateLimitExceeded', 'rateLimitExceeded', ):
            return True
    # Everything else is a success or non-retriable so break.
    return False
```

I'm not sure if we should comment in a more detailed way here, given that the logic is beyond our control. Do you think we should make the comment more informative or maybe link to the `http.py`'s source code?

Reply:
Nope, I think this is fine as-is.